### ISSUE TITLE
Possible to specify phylogenetic covariance / PGLS structures in glmmTMB?

### ISSUE THREAD

**Issue by florianhartig** [glmmTMB/glmmTMB#685]
Hi glmmTMB crew,

is there any way to specify a PGLS-type covariance structure in glmmTMB, akin to https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/corBrownian? 

If not, I think it would be great if that could be supported - I just finished a course on teaching GLMMs for ecologists, and the phylogenetic regression was something that people were really interested in, but when moving beyond gls, the R ecosystem becomes quite fragmented. It would be great to have the ability to fit a basic PGLS in glmmTMB to be able to combine PGLS with all the other options of glmmTMB. 

Best,
Florian

**Comment by mebrooks** [comment]
I've been thinking the exact same thing now that glmmTMB has more potential for joint species distribution models. 

I was looking into the animal model years ago and took notes that Ben made some progress and documented it here
https://github.com/bbolker/mixedmodels-misc/blob/master/notes/phylog.rmd

Also here's a thread that touched on the subject https://stat.ethz.ch/pipermail/r-sig-mixed-models/2017q4/026074.html

**Comment by bbolker** [comment]
There's more on this, when I get a chance to get back to it.  [This file](https://github.com/glmmTMB/glmmTMB/blob/master/misc/fixcorr.rmd) shows examples of working with a fixed correlation matrix, which is essentially all you need (although doing it this way seems very slow).  See also https://zenodo.org/record/2639887 , which works with glmmTMB and lme4 ...

**Comment by mebrooks** [comment]
Since [latent variables like in rr covstruct]( https://doi.org/10.1111/2041-210X.13303) can absorb residual correlations between species, I'm wondering if it might be ok (or when) to leave out an explicit phylogenetic correlation.

**Comment by florianhartig** [comment]
I don't think this is a replacement - a jSDM such as gllvm fits an unconstrained covariance, i.e. for N species N*(N-1)/2 parameters, and even with a low-rank approximation such as the latent structure k*N, k being the latents. 

The phylogenetic residual structures impose a fixed structure on the residuals, with zero to a few parameters. It's quite different imo.

**Comment by Logos-Lu** [comment]
I would like to inquire about any recent developments regarding the implementation of PGLS-type covariance structure in glmmTMB, following up on @florianhartig 's question. This information would be highly valuable not only to me but also to ecologists who are considering transitioning from gls to glmmTMB.

**Comment by vcgarnica** [comment]
Hi everyone,
I'm encountering an error when trying to run the tutorial: Error: could not find function "propto".
Does anyone know which package contains this function or is no longer maintained?

**Comment by bbolker** [comment]
Can you please be more explicit about what you mean by "the tutorial", i.e. give a URL? (There are several documents linked in the thread above, it's hard to know which one you mean).

Without knowing more about what you're doing: the `propto` covariance structure was added in version 1.1.10 of the package, so you should make sure you have installed a version at least that recent ...

**Comment by vcgarnica** [comment]
Right on, I had the old version. Thank you for your help and contribution

**Comment by bbolker** [comment]
I think I'm going to close this now that we *do* have the `propto` covariance structure (although I can imagine ways to make it more convenient, i.e. `phylo()` structure that automatically translates a phylogenetic tree to a covariance matrix internally ...)

**Comment by bbolker** [comment]
PS it might be worth suggesting to the author/maintainer of whatever document you're looking at that they make this dependence (`glmmTMB` version >= 1.1.10) explicit in the text ...
